From 64b290e98889e8d5b93036195539ab7a8cca92f9 Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:40:31 -0400
Subject: [alsa] hda/realtek: Fixed ALC286/ALC288 recording delay for Headset Mic

Message-id: <140972823573-412-patchset-jkysela@redhat.com>
Patchwork-id: 91108
O-Subject: [RHEL 7.1 PATCH 411/442] ALSA: hda/realtek - Fixed ALC286/ALC288 recording delay for Headset Mic
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

It will be recording voice delay for resume back recording for Headset Mic.
This alc286 will quickly open Headset Mic, to prevent avoid recording files are missing.
The issue was fixed. This is follow ALC286 programing guide.

[fix build error, add static and renamed the function by tiwai]

Signed-off-by: Kailang Yang <kailang@realtek.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: f7ae9ba0e80dd3a359cd9e4075b7b6f8acc5a3fa)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/pci/hda/patch_realtek.c b/sound/pci/hda/patch_realtek.c
index 654c8f1..69f2744 100644
--- a/sound/pci/hda/patch_realtek.c
+++ b/sound/pci/hda/patch_realtek.c
@@ -2782,6 +2782,27 @@ static int alc269_parse_auto_config(struct hda_codec *codec)
  return alc_parse_auto_config(codec, alc269_ignore, ssids);
 }
 
+static int find_ext_mic_pin(struct hda_codec *codec);
+
+static void alc286_shutup(struct hda_codec *codec)
+{
+ int i;
+ int mic_pin = find_ext_mic_pin(codec);
+ /* don't shut up pins when unloading the driver; otherwise it breaks
+  * the default pin setup at the next load of the driver
+  */
+ if (codec->bus->shutdown)
+  return;
+ for (i = 0; i < codec->init_pins.used; i++) {
+  struct hda_pincfg *pin = snd_array_elem(&codec->init_pins, i);
+  /* use read here for syncing after issuing each verb */
+  if (pin->nid != mic_pin)
+   snd_hda_codec_read(codec, pin->nid, 0,
+     AC_VERB_SET_PIN_WIDGET_CONTROL, 0);
+ }
+ codec->pins_shutup = 1;
+}
+
 static void alc269vb_toggle_power_output(struct hda_codec *codec, int power_up)
 {
  int val = alc_read_coef_idx(codec, 0x04);
@@ -5384,6 +5405,7 @@ static int patch_alc269(struct hda_codec *codec)
  case 0x10ec0286:
  case 0x10ec0288:
   spec->codec_variant = ALC269_TYPE_ALC286;
+  spec->shutup = alc286_shutup;
   break;
  case 0x10ec0255:
   spec->codec_variant = ALC269_TYPE_ALC255;
-- 
1.7.1